### ISSUE TITLE
Get-AzureRmRoleAssignment command failing

### ISSUE THREAD

**Issue by dmittal02** [Azure/azure-powershell#3407]
### PowerShell Version
Instructions: to get PowerShell version, type `$PSVersionTable` and look for the value associated with `PSVersion`
latest
### Module Version
Azure PS version: 1.0.0
### OS Version
Instructions: to get OS version, type `$PSversionTable` and look for value associated with `BuildVersion`
Windows 2012 R2
### Description
Error when trying to run the PS command:
Get-AzureRmRoleAssignment 
: ##[error]Exception of type 'Microsoft.Rest.Azure.CloudException' was thrown.
2017-01-18T08:40:40.1990848Z Microsoft.Azure.Commands.Resources.Models.Authorization.PSRoleDefinition
### Debug Output
Instructions: to get Debug Output, set `$DebugPreference="Continue"` and then execute the cmdlet or script causing the issue

### Script/Steps for Reproduction
2017-01-18T17:43:54.3907947Z ##[debug]Leaving Initialize-Azure.
2017-01-18T17:43:54.4499767Z ##[command]& 'C:\Users\buildguest\AppData\Local\Temp\bfdd2b32-1828-48c5-a451-d7143f0a9e30.ps1' 
2017-01-18T17:43:56.0065186Z ##[debug]Error record:
2017-01-18T17:43:56.1317540Z ##[debug]Get-AzureRmRoleAssignment : Exception of type 'Microsoft.Rest.Azure.CloudException' was thrown.
2017-01-18T17:43:56.1337554Z ##[debug]At C:\Users\buildguest\AppData\Local\Temp\bfdd2b32-1828-48c5-a451-d7143f0a9e30.ps1:3 char:1
2017-01-18T17:43:56.1357540Z ##[debug]+ Get-AzureRmRoleAssignment -verbose
2017-01-18T17:43:56.1367563Z ##[debug]+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
2017-01-18T17:43:56.1387550Z ##[debug]    + CategoryInfo          : CloseError: (:) [Get-AzureRmRoleAssignment], CloudException
2017-01-18T17:43:56.1407543Z ##[debug]    + FullyQualifiedErrorId : Microsoft.Azure.Commands.Resources.GetAzureRoleAssignmentCommand
2017-01-18T17:43:56.1417557Z ##[debug] 
2017-01-18T17:43:56.1447543Z ##[debug]Script stack trace:
2017-01-18T17:43:56.1497562Z ##[debug]at <ScriptBlock>, C:\Users\buildguest\AppData\Local\Temp\bfdd2b32-1828-48c5-a451-d7143f0a9e30.ps1: line 3
2017-01-18T17:43:56.1517543Z ##[debug]at <ScriptBlock>, <No file>: line 1
2017-01-18T17:43:56.1527562Z ##[debug]at <ScriptBlock>, C:\a\_tasks\AzurePowerShell_72a1931b-effb-4d2e-8fd8-f8472a07cb62\1.1.5\AzurePowerShell.ps1: line 64
2017-01-18T17:43:56.1547543Z ##[debug]at <ScriptBlock>, C:\a\_tasks\AzurePowerShell_72a1931b-effb-4d2e-8fd8-f8472a07cb62\1.1.5\AzurePowerShell.ps1: line 60
2017-01-18T17:43:56.1557558Z ##[debug]at <ScriptBlock>, <No file>: line 1
2017-01-18T17:43:56.1577601Z ##[debug]at <ScriptBlock>, <No file>: line 22
2017-01-18T17:43:56.1597546Z ##[debug]at <ScriptBlock>, <No file>: line 18
2017-01-18T17:43:56.1613921Z ##[debug]at <ScriptBlock>, <No file>: line 1
2017-01-18T17:43:56.1644002Z ##[debug]Exception:
2017-01-18T17:43:56.2355650Z ##[debug]Microsoft.Rest.Azure.CloudException: Exception of type 'Microsoft.Rest.Azure.CloudException' was thrown.
2017-01-18T17:43:56.2375685Z ##[debug]   at Microsoft.Azure.Graph.RBAC.ObjectsOperations.<GetObjectsByObjectIdsWithHttpMessagesAsync>d__6.MoveNext()
2017-01-18T17:43:56.2395630Z ##[debug]--- End of stack trace from previous location where exception was thrown ---
2017-01-18T17:43:56.2415644Z ##[debug]   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
2017-01-18T17:43:56.2425679Z ##[debug]   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
2017-01-18T17:43:56.2445651Z ##[debug]   at Microsoft.Azure.Graph.RBAC.ObjectsOperationsExtensions.<GetObjectsByObjectIdsAsync>d__3.MoveNext()
2017-01-18T17:43:56.2465644Z ##[debug]--- End of stack trace from previous location where exception was thrown ---
2017-01-18T17:43:56.2485657Z ##[debug]   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
2017-01-18T17:43:56.2495653Z ##[debug]   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
2017-01-18T17:43:56.2516192Z ##[debug]   at Microsoft.Azure.Graph.RBAC.ObjectsOperationsExtensions.GetObjectsByObjectIds(IObjectsOperations operations, GetObjectsParameters parameters)
2017-01-18T17:43:56.2536390Z ##[debug]   at Microsoft.Azure.Commands.Resources.Models.ActiveDirectory.ActiveDirectoryClient.GetObjectsByObjectId(List`1 objectIds)
2017-01-18T17:43:56.2549680Z ##[debug]   at Microsoft.Azure.Commands.Resources.Models.Authorization.AuthorizationClientExtensions.ToPSRoleAssignments(IEnumerable`1 assignments, List`1 roleDefinitions, AuthorizationClient policyClient, ActiveDirectoryClient activeDirectoryClient, Boolean excludeAssignmentsForDeletedPrincipals)
2017-01-18T17:43:56.2569844Z ##[debug]   at Microsoft.Azure.Commands.Resources.Models.Authorization.AuthorizationClient.FilterRoleAssignments(FilterRoleAssignmentsOptions options, String currentSubscription)
2017-01-18T17:43:56.2579843Z ##[debug]   at Microsoft.Azure.Commands.Resources.GetAzureRoleAssignmentCommand.ExecuteCmdlet()
2017-01-18T17:43:56.2599842Z ##[debug]   at Microsoft.WindowsAzure.Commands.Utilities.Common.AzurePSCmdlet.ProcessRecord()
2017-01-18T17:43:56.2959008Z ##[error]Exception of type 'Microsoft.Rest.Azure.CloudException' was thrown.
2017-01-18T17:43:56.2959008Z ##[debug]Processed: ##vso[task.logissue type=error]Exception of type 'Microsoft.Rest.Azure.CloudException' was thrown.
2017-01-18T17:43:56.2989006Z ##[debug]Processed: ##vso[task.complete result=Failed]
2017-01-18T17:43:56.3229004Z ##[section]Finishing: Azure PowerShell script: InlineScript

**Comment by markcowl** [comment]
@dmittal02 Please include the text of the cmdlet invocation (what command line did you use) and a description of the expected result.  Also, can you try your command with the latest version?

**Comment by dmittal02** [comment]
To give you more context, we are using this command line in VSTS build Azure PowerShell build task.
this is the command line, we use:
Get-AzureRmRoleAssignment
and this command line will be executed on hosted build agent.
In fact for customer when he run it on the private build agent to which he can remote, he get the required data, but I am getting authorization error there as well.
Please let me know if that helps. You can also reach out to me at: dmittal@microsoft.com to discuss in more detail.

**Comment by efridman** [comment]
@markcowl We get the same error when we try to call Get-AzureRmRoleAssignment in our VSTS release pipeline from an Azure PowerShell release task. This is running on a hosted agent with an Azure Resource Manager connection. The Azure PowerShell version on the hosted agent is 3.1.0.

**Comment by dmittal02** [comment]
@markcowl Here is the script as with which we are able to reproduce it locally:

$psCredential = New-Object System.Management.Automation.PSCredential(
            "serviceprincipalID",
            (ConvertTo-SecureString "Password" -AsPlainText -Force)) 
Add-AzureRMAccount -ServicePrincipal -Tenant "<tenantID>" -Credential $psCredential  
$additional = @{} 
$additional['TenantId'] = "<TenantID>" 
Select-AzureRMSubscription -SubscriptionId "<SubscriptionID>" @additional   

Get-AzureRmRoleAssignment -debug

You will get the issue. 
Service principal ID choosen is owner on the subscription.

**Comment by markcowl** [comment]
What are the permissions in ActiveDirectory of your service principal?  To use the role assignment cmdlets,  you need access to the graph api, which is managed by permissions granted to your service principal in AD.  To view role assignments, you will need at least read access to the tenant in AD.

**Comment by efridman** [comment]
@markcowl It looks like missing AAD permissions were the cause of our issue. Thanks.

**Comment by markcowl** [comment]
@efridman @dmittal02 Can you guys verify that this was resolved?

**Comment by efridman** [comment]
@markcowl Yes, our problem did turn out to be that the SPN did not have permissions on AAD. Thanks.

**Comment by keeshuizer** [comment]
@efridman We are running into the same problem. Can you elaborate on what permissions you gave the SPN? We have been trying things, but still no success

**Comment by efridman** [comment]
@keeshuizer We actually did not check this all the way through, since we don't want to give the SPN permissions on AAD for other reasons. But permissions were missing and that's why we thought that was the issue. I would assume read permissions to read roles and write permissions to assign them, but I'm guessing you tried that?

**Comment by markcowl** [comment]
I have a repro of the issue with current cmdlets, it is definitely related to the whitelist for graph api-version used in the call to graph to get friendly names for the users and groups, and not the permissions on the service principal.

Here is a request/response showing the failure
```
DEBUG: ============================ HTTP REQUEST ============================

HTTP Method:
POST

Absolute Uri:
https://graph.windows.net/54826b22-38d6-4fb2-bad9-b7b93a3e9c5a/getObjectsByObjectIds?api-version=1.6-internal

Headers:
x-ms-client-request-id        : 83854599-8e85-4858-95c9-db7eae89ad68
accept-language               : en-US

Body:
{
  "objectIds": [
    -- redacted --
  ],
  "includeDirectoryObjectReferences": true
}

DEBUG: ============================ HTTP RESPONSE ============================

Status Code:
Forbidden

Headers:
ocp-aad-diagnostics-server-name: XYNBldSbijB5bKKmRhxq1vFYS8rdTdBj0OJLIY4iTms=
request-id                    : 6bd997bd-a7f4-4692-ae08-f19f5136157a
client-request-id             : 097a0dcb-d072-4a10-a230-cd4e198547e2
x-ms-dirapi-data-contract-version: 1.6-internal
ocp-aad-session-key           :
ynVvvPIP2mAgozVdI3czQ2GZgS8yxN9FKz-wyhh3yVii1HCy93ZYZagfW5owHDfYGypn_eXgD9i1O6Wq6utrZj-1dkEXjoPQqXlFrB3Q36sZA2DqTI56p_-
cPXRiM6pN.FJsGF_JL39ZHV7R7XgD0DkGidkT-KaokgM4QPdcOdE0
Strict-Transport-Security     : max-age=31536000; includeSubDomains
Access-Control-Allow-Origin   : *
Duration                      : 632811
Cache-Control                 : private
Server                        : Microsoft-IIS/8.5
X-AspNet-Version              : 4.0.30319
X-Powered-By                  : ASP.NET,ASP.NET
Date                          : Thu, 23 Feb 2017 01:32:14 GMT

Body:
{
  "odata.error": {
    "code": "Authentication_Unauthorized",
    "message": {
      "lang": "en",
      "value": "Access denied to the specified API version."
    }
  }
}
```

**Comment by gitralf** [comment]
cmdlet nor working for me, too. A simple Get-AzureRMRoleAssignment ist not working any more and I don't have any idea what changed... Here is the output, pls. help:

PS C:\PublicProjects\PowerShell> **Login-AzureRmAccount**


Environment           : AzureCloud
Account               : admin@d...d.onmicrosoft.com
TenantId              : 8dff03-...-312cf7c
SubscriptionId        : abcde-730316a2e9e0
SubscriptionName      : Visual Studio Ultimate mit MSDN
CurrentStorageAccount :



PS C:\PublicProjects\PowerShell> **Get-AzureRmRoleAssignment**
Get-AzureRmRoleAssignment : Run Login-AzureRmAccount to login.
In Zeile:1 Zeichen:1
+ Get-AzureRmRoleAssignment
+ ~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation: (:) [Get-AzureRmRoleAssignment], PSInvalidOperationException
    + FullyQualifiedErrorId : InvalidOperation,Microsoft.Azure.Commands.Resources.GetAzureRoleAssignmentCommand

**Comment by markcowl** [comment]
@gitralf This looks unrelated to the issue at hand - the issue you are seeing is likely that you have incompatible versions of the cmdlets installed.  Normally you can resolve this by:

(1) Installing the latest MSI -or-
(2) Updating to the latest AzureRM and Azure modules from the PSGallery.

If you'd like to confirm this is the issue, run

```Get-Module -ListAvailable```

And send the output - most likely there will be some mismatched moduel versions.

**Comment by markcowl** [comment]
This issue is due to be resolved when the newest version of AAD Graph is released. This is due to heppen in the May release, so moving this to May

**Comment by bganapa** [comment]
@shuagarw  We need the fix for azurestack in the stack-dev branch

**Comment by twitchax** [comment]
@shuagarw, any update on this?

**Comment by shuagarw** [comment]
@rbacask for FYI
+ @darshanhs90 as he is working on this fix in master branch.  @afshins to port the fix in AzureStack once it is completed in master branch.

**Comment by markcowl** [comment]
This should be fixed.  Please let us know if you see a recurrence

**Comment by 4c74356b41** [comment]
appears to be working for me, thanks!

**Comment by YoannGUILLO** [comment]
Same error when I use the command with my Automation Account (Run As Account created by default with my Automation account)
Get-AzureRmRoleAssignment : Exception of type 'Microsoft.Rest.Azure.CloudException' was thrown.
My modules have been updated, AzureRM.Resources : version 4.1.0

**Comment by RBACAsk** [comment]
Can you share the exception details?

Thx
Surabhi

**Comment by darshanhs90** [comment]
@YoannGUILLO could you run the command after setting $DebugPreference="Continue"
so that it will be easier to know the cause for this issue

**Comment by RBACAsk** [comment]
Looks like the user or service principal that you are logged-in with doesn’t have sufficient permissions on AAD Graph side.

https://graph.windows.net/24139d14-c62c-4c47-8bdd-ce71ea1d50cf/getObjectsByObjectIds?api-version=1.6<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgraph.windows.net%2F24139d14-c62c-4c47-8bdd-ce71ea1d50cf%2FgetObjectsByObjectIds%3Fapi-version%3D1.6&data=02%7C01%7CSurabhi.Pandey%40microsoft.com%7Cf2f149c28e0c482341ec08d4b3272d82%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636330428246200728&sdata=8V9kHFJx7Eef7TxFJIVyT6YEuoCQ8vd1CAU58HCEKNc%3D&reserved=0>

Can you check the permissions?

Thx
Surabhi

**Comment by YoannGUILLO** [comment]
The account is owner on the subscription.

**Comment by RBACAsk** [comment]
I was referring to permissions in the AAD tenant. The permissions for calls to graph.windows.net are not mapped to Azure subscriptions.

Thx
Surabhi

**Comment by YoannGUILLO** [comment]
How can I check these permissions ? 
The run as service account is owner of my subscription, I don't know how to check the permissions to call graph.windows.net

**Comment by RBACAsk** [comment]
You can log into portal.azure.com<http://portal.azure.com> using an account from this tenant and then check the Applications in the Azure Active Directory section.

+ Vadim for more info.

Thx
Surabhi

_____________________________

**Comment by ondas** [comment]
@RBACAsk - What permissions are necessary for the application to have to get around this error? The webapp/API that's running the command is an owner of the resource group and I believe it has access to the Graph API, but I get the same "Microsoft.Rest.Azure.CloudException: Exception of type 'Microsoft.Rest.Azure.CloudException' was thrown at Microsoft.Azure.Graph.RBAC" error. Running on fully updated Powershell/AzurePS

**Comment by darshanhs90** [comment]
@ondas could you run the command with $DebugPreference="Continue" and let us know the exact issue that comes up in that.
The issue mentioned in this thread was fixed in azure powershell V3.9+

**Comment by ondas** [comment]
I ended up finding a workaround-- the command was working but throwing an error so I put it in a try block and it fixed it.

**Comment by javiermarasco** [comment]
I'm having the same issue using a user that's owner in the subscription I'm running the command. This was working just fine a couple of months ago.

**Comment by darshanhs90** [comment]
javiermarasco could you run the command after setting $DebugPreference="Continue" and provide us with the log, so that it will be easier to find out the issue causing this

**Comment by javiermarasco** [comment]
@darshanhs90 the command simply doesn't complete, I enabled the -debug flag in the command and only get this:
DEBUG: 05:17:28 p.m. - GetAzureRoleAssignmentCommand begin processing with ParameterSet 'EmptyParameterSet
'.
DEBUG: 05:17:30 p.m. - using account id 'user@comain.com'...

the account is edited, it shows my account.

**Comment by javiermarasco** [comment]
@darshanhs90 Any idea on what this could be? I'm completely lost here, this worked just fine a couple of months ago.

**Comment by darshanhs90** [comment]
@javiermarasco Could you use the latest azure powershell https://github.com/Azure/azure-powershell/releases/tag/v4.3.1-August2017 and let me know if that helps.
If that doesnt work ,you can provide me with the fiddler trace or debug output for me to dig deep into the issue.

**Comment by Steven-N** [comment]
Specifically, which permissions on the graph API does the service principal require in AAD to run the 'Get-AzureRmRoleAssignment' command? I am getting the same error as well on the most up to date version of Azure-Powershell.

**Comment by jakobli** [comment]
The odd thing for me is using New-AzureRoleAssignment works but I can't do Get-AzureRMRoleAssignment

**Comment by Steven-N** [comment]
@darshanhs90 Can you assist with this? I'm still getting the same error message. The SP is owner on the subscription. It has Directory R/W access on the Graph API and I still see the same error when it tries to access the Graph API (getObjectsByObjectIds).

> Body:
> {
>   "odata.error": {
>     "code": "Authorization_RequestDenied",
>     "message": {
>       "lang": "en",
>       "value": "Insufficient privileges to complete the operation."
>     }
>   }
> }

**Comment by darshanhs90** [comment]
@jakobli Could you share the stacktrace to help you out regarding the issue

**Comment by Dule011** [comment]
I'm running into the same issue and am fairly certain that it is related to lack of access on Graph API. Reading thru this thread i do not see what specific rights need to be granted. Steven-N and ondas have asked but i can't find the answer.

**Comment by Steven-N** [comment]
@Dule011 Give the Service Principal access on Active Directory to 'Read and Write on the Directory'. That should fix your issues.

**Comment by Dule011** [comment]
Thanks, got it working. For the record: "read" is sufficient, no need to have "write". @Steven-N, you're right, nothing to do with Graph, just read access to AAD fixed my issue.

**Comment by maikvandergaag** [comment]
Could this be reopened?

I still got the issue when running these comments in VSTS:
On the New-AzureRmRoleAssignment I receive a Microsoft.Rest.Azure.CloudException also on the Get method.

**Comment by darshanhs90** [comment]
@maikvandergaag could you run the command with -Debug and send the stack trace

**Comment by adhamod** [comment]
Same error when I use the command with my Automation Account (Run As Account created by default with my Automation account). I also went to the permissions under App Registration and added Microsoft Graph and Windows Azure Active Directory. 
Get-AzureRmRoleAssignment : Exception of type 'Microsoft.Rest.Azure.CloudException' was thrown.
My modules have been updated, AzureRM.Resources : version 5.5.1.

Here is the output:
ServerMessage    : :  (System.Collections.Generic.List`1[Microsoft.Rest.Azure.CloudError])

ServerResponse   : {Forbidden}

RequestMessage   : {POST 

                   https://graph.windows.net/72f988bf-86f1-41af-91ab-2d7cd011db47/getObjectsByObjectIds?api-version=1.6}

RequestId        : 

InnerException   : False

Exception        : Microsoft.Rest.Azure.CloudException: Exception of type 'Microsoft.Rest.Azure.CloudException' was 

                   thrown.

                      at Microsoft.Azure.Graph.RBAC.Version1_6.ObjectsOperations.<GetObjectsByObjectIdsWithHttpMessagesA

                   sync>d__6.MoveNext()

                   --- End of stack trace from previous location where exception was thrown ---

                      at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()

                      at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)

                      at Microsoft.Azure.Graph.RBAC.Version1_6.ObjectsOperationsExtensions.<GetObjectsByObjectIdsAsync>d

                   __3.MoveNext()

                   --- End of stack trace from previous location where exception was thrown ---

                      at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()

                      at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)

                      at Microsoft.Azure.Graph.RBAC.Version1_6.ActiveDirectory.ActiveDirectoryClient.GetObjectsByObjectI

                   d(List`1 objectIds)

                      at Microsoft.Azure.Commands.Resources.Models.Authorization.AuthorizationClientExtensions.ToPSRoleA

                   ssignments(IEnumerable`1 assignments, List`1 roleDefinitions, AuthorizationClient policyClient, 

                   ActiveDirectoryClient activeDirectoryClient, Boolean excludeAssignmentsForDeletedPrincipals)

                      at Microsoft.Azure.Commands.Resources.Models.Authorization.AuthorizationClientExtensions.ToPSRoleA

                   ssignments(IEnumerable`1 assignments, AuthorizationClient policyClient, ActiveDirectoryClient 

                   activeDirectoryClient, String scopeForRoleDefinitions, Boolean 

                   excludeAssignmentsForDeletedPrincipals)

                      at Microsoft.Azure.Commands.Resources.Models.Authorization.AuthorizationClient.FilterRoleAssignmen

                   ts(FilterRoleAssignmentsOptions options, String currentSubscription)

                      at Microsoft.Azure.Commands.Resources.GetAzureRoleAssignmentCommand.ExecuteCmdlet()

                      at Microsoft.WindowsAzure.Commands.Utilities.Common.AzurePSCmdlet.ProcessRecord()

Message          : Exception of type 'Microsoft.Rest.Azure.CloudException' was thrown.

StackTrace       :    at Microsoft.Azure.Graph.RBAC.Version1_6.ObjectsOperations.<GetObjectsByObjectIdsWithHttpMessagesA

                   sync>d__6.MoveNext()

                   --- End of stack trace from previous location where exception was thrown ---

                      at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()

                      at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)

                      at Microsoft.Azure.Graph.RBAC.Version1_6.ObjectsOperationsExtensions.<GetObjectsByObjectIdsAsync>d

                   __3.MoveNext()

                   --- End of stack trace from previous location where exception was thrown ---

                      at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()

                      at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)

                      at Microsoft.Azure.Graph.RBAC.Version1_6.ActiveDirectory.ActiveDirectoryClient.GetObjectsByObjectI

                   d(List`1 objectIds)

                      at Microsoft.Azure.Commands.Resources.Models.Authorization.AuthorizationClientExtensions.ToPSRoleA

                   ssignments(IEnumerable`1 assignments, List`1 roleDefinitions, AuthorizationClient policyClient, 

                   ActiveDirectoryClient activeDirectoryClient, Boolean excludeAssignmentsForDeletedPrincipals)

                      at Microsoft.Azure.Commands.Resources.Models.Authorization.AuthorizationClientExtensions.ToPSRoleA

                   ssignments(IEnumerable`1 assignments, AuthorizationClient policyClient, ActiveDirectoryClient 

                   activeDirectoryClient, String scopeForRoleDefinitions, Boolean 

                   excludeAssignmentsForDeletedPrincipals)

                      at Microsoft.Azure.Commands.Resources.Models.Authorization.AuthorizationClient.FilterRoleAssignmen

                   ts(FilterRoleAssignmentsOptions options, String currentSubscription)

                      at Microsoft.Azure.Commands.Resources.GetAzureRoleAssignmentCommand.ExecuteCmdlet()

                      at Microsoft.WindowsAzure.Commands.Utilities.Common.AzurePSCmdlet.ProcessRecord()

HelpLink         : 

ErrorDetails     : 

ErrorCategory    : CloseError: (:) [Get-AzureRmRoleAssignment], CloudException

InvocationInfo   : System.Management.Automation.InvocationInfo

ScriptStackTrace : at <ScriptBlock>, <No file>: line 46

**Comment by uday31in** [comment]
'''
DEBUG: ============================ HTTP REQUEST ============================

HTTP Method:
POST

Absolute Uri:
https://graph.windows.net/<aad-tenant-id>/getObjectsByObjectIds?api-version=1.6

Headers:
x-ms-client-request-id        : cdf7915c-f189-4f50-bf87-307769971872
accept-language               : en-US

Body:
{
  "objectIds": [
    "c15807e8-e800-4aab-bf3c-2d7b42ec915b"
  ],
  "includeDirectoryObjectReferences": true
}

DEBUG: ============================ HTTP RESPONSE ============================

Status Code:
Forbidden

Headers:
Pragma                        : no-cache
ocp-aad-diagnostics-server-name: gVYeWZfSX9mkzriipUjAD9i6IZAJ5eI2PiGNN2hB0oU=
request-id                    : a9821b91-8e25-4e21-b9c8-572783772bd0
client-request-id             : 1a1cc7fe-0545-4449-838e-ed3a0e141da7
x-ms-dirapi-data-contract-version: 1.6
ocp-aad-session-key           : FHsP3AaS_pcxlR13h7VnmUweSN6RNfikqxGLCCCF9ysLri0exLgNWn50
X-Content-Type-Options        : nosniff
DataServiceVersion            : 3.0;
Strict-Transport-Security     : max-age=31536000; includeSubDomains
Access-Control-Allow-Origin   : *
Duration                      : 2965843
Cache-Control                 : no-cache
Server                        : Microsoft-IIS/10.0
X-AspNet-Version              : 4.0.30319
X-Powered-By                  : ASP.NET,ASP.NET
Date                          : Wed, 02 May 2018 21:26:37 GMT

Body:
{
  "odata.error": {
    "code": "Authorization_RequestDenied",
    "message": {
      "lang": "en",
      "value": "Insufficient privileges to complete the operation."
    }
  }
}
'''



'''
RequestId      :
Message        : Exception of type 'Microsoft.Rest.Azure.CloudException' was thrown.
ServerMessage  : :  (System.Collections.Generic.List`1[Microsoft.Rest.Azure.CloudError])
ServerResponse : {Forbidden}
RequestMessage : {POST https://graph.windows.net/virtualdatacenteroutlook.onmicrosoft.com/getObjectsByObjectIds?api-ver
                 sion=1.6}
InvocationInfo : {Get-AzureRmRoleAssignment}
Line           : Get-AzureRmRoleAssignment -Scope /subscriptions/15e50875-e180-4d16-a8cb-57361c6a3cfb -ObjectId
                 c15807e8-e800-4aab-bf3c-2d7b42ec915b  -RoleDefinitionId  8e3af657-a8ff-443c-a75c-2fe8c4bcb635
Position       : At line:1 char:1
                 + Get-AzureRmRoleAssignment -Scope /subscriptions/15e50875-e180-4d16-a8 ...
                 + ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
StackTrace     :    at Microsoft.Azure.Graph.RBAC.Version1_6.ObjectsOperations.<GetObjectsByObjectIdsWithHttpMessagesAs
                 ync>d__6.MoveNext()
                 --- End of stack trace from previous location where exception was thrown ---
                    at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
                    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
                    at Microsoft.Azure.Graph.RBAC.Version1_6.ObjectsOperationsExtensions.<GetObjectsByObjectIdsAsync>d_
                 _3.MoveNext()
                 --- End of stack trace from previous location where exception was thrown ---
                    at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
                    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
                    at Microsoft.Azure.Graph.RBAC.Version1_6.ObjectsOperationsExtensions.GetObjectsByObjectIds(IObjects
                 Operations operations, GetObjectsParameters parameters)
                    at Microsoft.Azure.Graph.RBAC.Version1_6.ActiveDirectory.ActiveDirectoryClient.GetObjectsByObjectId
                 (List`1 objectIds)
                    at Microsoft.Azure.Commands.Resources.Models.Authorization.AuthorizationClientExtensions.ToPSRoleAs
                 signments(IEnumerable`1 assignments, List`1 roleDefinitions, AuthorizationClient policyClient,
                 ActiveDirectoryClient activeDirectoryClient, Boolean excludeAssignmentsForDeletedPrincipals)
                    at Microsoft.Azure.Commands.Resources.Models.Authorization.AuthorizationClient.FilterRoleAssignment
                 s(FilterRoleAssignmentsOptions options, String currentSubscription)
                    at Microsoft.Azure.Commands.Resources.GetAzureRoleAssignmentCommand.ExecuteCmdlet()
                    at Microsoft.WindowsAzure.Commands.Utilities.Common.AzurePSCmdlet.ProcessRecord()
HistoryId      : 6

'''

**Comment by prak2419** [comment]
Ran into the same issue today using the latest AZ.Resources module. 

```
PS C:\WINDOWS\system32> $PSversionTable.PSversion

Major  Minor  Build  Revision
-----  -----  -----  --------
5      1      18362  145
```
![image](https://user-images.githubusercontent.com/41781768/68936180-052a6080-07c0-11ea-8620-41b5375aa3b1.png)


The service principal has all rights to AAD and Graph API. The issue is encountered while using a release pipeline, Powershell and an automation runbook. Get-AZRoleDefinition and New-AZRoleDefinition along with New-AZRoleAssignment is working but Get-AZRoleAssignment is failing. I have also added custom RBAC specifically for the roles to the SP,

> 
>              "Microsoft.Management/managementGroups/read",
> 		"Microsoft.Authorization/roleAssignments/*",
> 		"Microsoft.Authorization/roleDefinitions/*",
> 		"Microsoft.Authorization/policyAssignments/*",
> 		"Microsoft.Authorization/policyDefinitions/*",
> 		"Microsoft.Authorization/policySetDefinitions/*"

 Below is the error message,

```
DEBUG: ============================ HTTP REQUEST ============================

HTTP Method:
POST

Absolute Uri:
x-ms-client-request-id        : 48e5149b-4d8b-4b14-b15c-f8ff73584eb3
accept-language               : en-US

Body:
{
  "objectIds": [
    "XXXXXXX",
  ],
  "includeDirectoryObjectReferences": true
}

DEBUG: ============================ HTTP RESPONSE ============================

Status Code:
Forbidden

Headers:
Pragma                        : no-cache
ocp-aad-diagnostics-server-name: mWVRrE/lLiIXk/+7MrU6NqGT7R5TaoyUNVdj9aKPxEo=
request-id                    : 3c76e3f5-d916-4c4e-932e-d373643b8ca9
client-request-id             : ca0f1cde-76e3-4e2a-8237-aea77039159e
x-ms-dirapi-data-contract-version: 1.6
ocp-aad-session-key           : 6_Gg8CZrpG7M4ZUe7TypCuqOZ_uCSuw_8dtxUhTIqdVzcXYOBQWZyTny-cc0G5H-w_KJUpgCIM61w1OL77DZu-SEnlV3dh94dRmgrzJ8t_oGKmI_4xG9OsbV8Fov8MaY.fBbBl9DP5Th8I8yfD31hLYkfFrCqrDZcd3LMQfTZtbI
DataServiceVersion            : 3.0;
Strict-Transport-Security     : max-age=31536000; includeSubDomains
Access-Control-Allow-Origin   : *
Duration                      : 1938952
Cache-Control                 : no-cache
Date                          : Fri, 15 Nov 2019 10:11:18 GMT
X-AspNet-Version              : 4.0.30319
X-Powered-By                  : ASP.NET

Body:
{
  "odata.error": {
    "code": "Authorization_RequestDenied",
    "message": {
      "lang": "en",
      "value": "Insufficient privileges to complete the operation."
    },
    "requestId": "3c76e3f5-d916-4c4e-932e-d373643b8ca9",
    "date": "2019-11-15T10:11:18"
  }
}
```
Is this a regression?

**Comment by davejhahn** [comment]
I see this was closed, what was the resolution?  I have an application registration that has Directory.ReadAll API permissions admin granted and I receive "Authorization_RequestDenied" when I use New-AzRoleAssignment with -debug parameter.  BUT, it creates the role.  

I've verified this.  Deleted the role through the portal, run, get the exception, and it is there after.

However, it does not seem to create it EVERY time.  Someone mentioned adding a try/catch, but if it isn't guaranteed to work, that won't work.  

So my SP has permissions identified as required in this thread, Directory.ReadAll, it still throws an exception, but works sometimes.
